pkg/streams/streams.go

```diff
@@ -0,0 +1,108 @@
+package streams
+
+import (
+	"context"
+	"sync"
+
+	"github.com/aquasecurity/tracee/types/trace"
+)
+
+// Stream is a stream of events
+type Stream struct {
+	// policy mask is a bitmap of policies that this stream is interested in
+	policyMask uint64
+	// events is a channel that is used to receive events from the stream
+	events chan trace.Event
+}
+
+// ReceiveEvents returns a channel that can be used to receive events from the stream
+func (s *Stream) ReceiveEvents() <-chan trace.Event {
+	return s.events
+}
+
+// Publish publishes an event to the stream,
+// but first check if this stream is interested in this event,
+// by checking the event's policy mask against the stream's policy mask.
+func (s *Stream) publish(ctx context.Context, event trace.Event) {
+	if s.shouldIgnorePolicy(event) {
+		return
+	}
+
+	// currently we are blocking if the channel is full,
+	// though I think we should instead drop events, because if you have
+	// two streams, one should not stop the other from receiving events,
+	// if they are not fast enough. But there also use cases where blocking
+	// might be favor, so th idea is to later make this configurable.
+	// TODO: allow this to be configurable (drop/block) (josedonizetti)
+	select {
+	case s.events <- event:
+	case <-ctx.Done():
+		return
+	}
+}
+
+// shouldIgnorePolicy checks if the stream should ignore the event
+func (s *Stream) shouldIgnorePolicy(event trace.Event) bool {
+	return s.policyMask&event.MatchedPoliciesUser == 0
+}
+
+// NewStream creates a new stream
+func NewStream(policyMask uint64, chanSize int) *Stream {
+	return &Stream{
+		policyMask: policyMask,
+		events:     make(chan trace.Event, chanSize),
+	}
+}
+
+// StreamManager manages streams
+type StreamsManager struct {
+	mutex       sync.Mutex
+	subscribers map[*Stream]struct{}
+}
+
+// NewStreamManager creates a new stream manager
+func NewStreamsManager() *StreamsManager {
+	return &StreamsManager{
+		mutex:       sync.Mutex{},
+		subscribers: make(map[*Stream]struct{}),
+	}
+}
+
+// Subscribe adds a stream to the manager
+func (sm *StreamsManager) Subscribe(stream *Stream) {
+	sm.mutex.Lock()
+	defer sm.mutex.Unlock()
+	sm.subscribers[stream] = struct{}{}
+}
+
+// Unsubscribe removes a stream from the manager
+func (sm *StreamsManager) Unsubscribe(stream *Stream) {
+	sm.mutex.Lock()
+	defer sm.mutex.Unlock()
+
+	// check if stream is subscribed
+	if _, ok := sm.subscribers[stream]; ok {
+		delete(sm.subscribers, stream)
+		close(stream.events)
+	}
+}
+
+// Publish publishes an event to all streams
+func (sm *StreamsManager) Publish(ctx context.Context, event trace.Event) {
+	sm.mutex.Lock()
+	defer sm.mutex.Unlock()
+
+	for stream := range sm.subscribers {
+		stream.publish(ctx, event)
+	}
```

Comment:
This loop can block, right? How about each stream having its publish() triggered in its own goroutine?

Reply:
Can you explain further? Because it doesn't seem like a benefit for me, blocking will always happen at some point because we are not using the unblocking  semantics to send events. Adding more channels/goroutines will not change it, blocking will still happen, and when all channel blocks this pressure returns to the pipeline. That is why there are comments about dropping events and using a queue per stream.

Reply:
> That is why there are comments about dropping events and using a queue per stream.

This is the right path already discussed 👍🏼. Ignore my comment above.